ntlm.go

```diff
@@ -23,93 +24,118 @@ import (
 // dialContext := (&net.Dialer{KeepAlive: 30*time.Second, Timeout: 30*time.Second}).DialContext
 type DialContext func(ctx context.Context, network, addr string) (net.Conn, error)
 
-// WrapDialContext wraps a DialContext with an NTLM Authentication to a proxy.
-func WrapDialContext(dialContext DialContext, proxyAddress, proxyUsername, proxyPassword, proxyDomain string) DialContext {
+// NewNTLMProxyDialContext provides a DialContext function that includes transparent NTLM proxy authentication.
+// Unlike WrapDialContext, it describes the proxy location with a full URL, whose scheme can be HTTP or HTTPS.
+func NewNTLMProxyDialContext(dialer *net.Dialer, proxyURL url.URL, proxyUsername, proxyPassword, proxyDomain string, tlsConfig *tls.Config) DialContext {
+	if dialer == nil {
+		dialer = &net.Dialer{}
+	}
 	return func(ctx context.Context, network, addr string) (net.Conn, error) {
-		conn, err := dialContext(ctx, network, proxyAddress)
```

Comment:
All this logic has been factored out into `dialAndNegotiate`, with no changes except the first line.